File: leetcode/2610/solution.py
"""
https://leetcode.com/problems/convert-an-array-into-a-2d-array-with-conditions/
2610. Convert an Array Into a 2D Array With Conditions
"""
from typing import List
from collections import Counter


class Solution:
    def findMatrix(self, nums: List[int]) -> List[List[int]]:
        arr = []
        counter = Counter(nums)
        count = sum(counter.values())
        while count > 0:
            sub_arr = []
            for k, v in counter.items():
                if v > 0:
                    sub_arr.append(k)
                    counter[k] -= 1
                    count -= 1
            arr.append(sub_arr)
        return arr

File: leetcode/2610/solution.rs
/**
 * https://leetcode.com/problems/convert-an-array-into-a-2d-array-with-conditions/
 * 2610. Convert an Array into a 2D Array with Conditions
 */
use std::collections::HashMap;

impl Solution {
    pub fn find_matrix(nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut counter: HashMap<i32, usize> = HashMap::new();
        nums.into_iter()
            .for_each(|n| *counter.entry(n).or_default() += 1);

        let mut max_val = counter.values().fold(0, |acc, cur| acc.max(*cur));
        let mut arr = vec![vec![]; max_val];
        for (n, count) in counter {
            for i in (0..count) {
                arr[i].push(n);
            }
        }

        arr
    }
}
